aws/rust-runtime/aws-http/src/content_encoding.rs

```diff
@@ -143,44 +137,57 @@ fn get_unsigned_chunk_bytes_length(payload_length: u64) -> u64 {
     if let Some(trailer_map) = trailer_map {
-        for (header_name, header_value) in trailer_map.into_iter() {
-            match header_name {
-                // New name, new value
-                Some(header_name) => {
-                    if already_wrote_first_trailer {
-                        // First trailer shouldn't have a preceding CRLF, but every trailer after it should
-                        trailers.write_str(CRLF).unwrap();
-                    } else {
-                        already_wrote_first_trailer = true;
-                    }
+        let mut current_header_name = None;
+        let mut trailers = BytesMut::new();
 
-                    trailers.write_str(header_name.as_str()).unwrap();
-                    trailers.write_char(':').unwrap();
-                }
-                // Same name, new value
-                None => {
-                    trailers.write_char(',').unwrap();
-                }
+        for (header_name, header_value) in trailer_map.into_iter() {
+            // When a header has multiple values, the name only comes up in iteration the first time
+            // we see it. Therefore, we need to keep track of the last name we saw and fall back to
+            // it when `header_name == None`.
+            current_header_name = header_name.or(current_header_name);
+
+            // In practice, this will always exist, but `if let` is nicer than unwrap
+            if let Some(header_name) = current_header_name.as_ref() {
+                trailers.extend_from_slice(header_name.as_ref());
+                trailers.extend_from_slice(TRAILER_SEPARATOR);
+                trailers.extend_from_slice(header_value.as_bytes());
+                trailers.extend_from_slice(CRLF.as_bytes());
             }
-            trailers.write_str(header_value.to_str().unwrap()).unwrap();
         }
-    }
 
-    // Write CRLF to end the body
-    trailers.write_str(CRLF).unwrap();
-    // If we wrote at least one trailer, we need to write an extra CRLF
-    if already_wrote_first_trailer {
-        trailers.write_str(CRLF).unwrap();
+        trailers
+    } else {
+        BytesMut::new()
     }
+}
 
-    trailers.into()
+/// Given an optional `HeaderMap`, calculate the total number of bytes required to represent the
+/// `HeaderMap`. If no `HeaderMap` is given as input, return 0.
+///
+/// - Trailer names are separated by a single colon only, no space.
+/// - Trailer names with multiple values will be written out one line per value, with the name
+///   appearing on each line.
+fn total_rendered_length_of_trailers(trailer_map: Option<&HeaderMap>) -> u64 {
+    match trailer_map {
+        Some(trailer_map) => trailer_map
+            .iter()
+            .map(|(trailer_name, trailer_value)| {
+                trailer_name.as_str().len()
+                    + TRAILER_SEPARATOR.len()
+                    + trailer_value.to_str().unwrap().len()
```

Comment:
The length header the trailer length gets incorporated into is in bytes, not characters, right? So the `to_str().unwrap()` is actually unnecessary and you can just count the bytes?